org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/classpath/ResolveClassPathsHandler.java

```diff
@@ -0,0 +1,153 @@
+
+import static java.util.Collections.emptySet;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.eclipse.core.resources.IContainer;
+import org.eclipse.core.resources.IFolder;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.IWorkspaceRoot;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+
+/**
+ * Class for resolving class path and getting location of the output directory for a java project.
+ */
+public class ResolveClassPathsHandler {
+
+  /**
+   * Resolves class path for a java project.
+   *
+   * @param arguments a list contains project URI
+   * @return the class paths entries
+   */
+  public static Set<String> resolveClasspaths(List<Object> arguments) {
+    String projectUri = (String) arguments.get(0);
+
+    IJavaProject javaProject = getJavaProject(projectUri);
+
+    if (javaProject == null) {
+      return emptySet();
+    }
+
+    return getProjectClassPath(javaProject);
+  }
+
+  /**
+   * Gets output location for a java project.
+   *
+   * @param arguments a list contains project URI
+   * @return output location, might returns empty string if something happens
+   */
+  public static String getOutputDirectory(List<Object> arguments) {
+    String projectUri = (String) arguments.get(0);
+
+    IJavaProject javaProject = getJavaProject(projectUri);
+
+    if (javaProject == null) {
+      return "";
+    }
+
+    try {
+      IPath outputLocation = javaProject.getOutputLocation();
+      IPath projectPath = javaProject.getProject().getFullPath();
+      IPath outputLocationWithoutProjectName =
+          outputLocation.removeFirstSegments(projectPath.segmentCount());
+      IFolder outputFolder = javaProject.getProject().getFolder(outputLocationWithoutProjectName);
+      if (outputFolder.exists()) {
+        return outputFolder.getLocation().toOSString();
+      }
+      return "";
+    } catch (JavaModelException e) {
+      return "";
+    }
+  }
+
+  private static IJavaProject getJavaProject(String projectUri) {
+    IWorkspaceRoot root = ResourcesPlugin.getWorkspace().getRoot();
+    IContainer[] containers = root.findContainersForLocationURI(JDTUtils.toURI(projectUri));
+
+    if (containers.length == 0) {
+      return null;
+    }
+
+    IContainer container = containers[0];
+    IProject project = container.getProject();
+    if (!project.exists()) {
+      return null;
+    }
+
+    return JavaCore.create(project);
+  }
+
+  /**
+   * Builds classpath for the java project.
+   *
+   * @param javaProject java project
+   * @return set of resources which are included to the classpath
+   */
+  private static Set<String> getProjectClassPath(IJavaProject javaProject) {
```

Comment:
This set contains paths to the resolved classpath entries. It can be a path to the library (folder or jar) or to the folder containing package fragments with source code. So I don't see why I can't use a set here. Maybe I'm missing something.

Reply:
Because you may have a classfile with the same fully qualified name in two different classpath entries. Which class file gets loaded depends on the order of classpath entries at runtime. Using a Set, the ordering of the classpath entries is lost.

Reply:
changed to list